### ISSUE TITLE
Allow bears to pass their own applicable actions

### ISSUE THREAD

**Issue by Makman2** [coala/coala#3650]
I would suggest to introduce a new system for results so bears can easily add possible fixes / choices for the user.

For example for the `GitCommitBear`: When a commit message does not meet the standards, the user should have the choice to amend the current commit (see https://github.com/coala/coala-bears/issues/1360).

To allow such functionality, I would propose a field inside `Result` which is a list of `ResultAction`-instances.
```
Result.actions = XYZAction(...)
```
And the bear would generate a result with the possible actions. It's possible to set even multiple ones:
```
my_result = Result('Something went wrong')
my_result.actions += ApplyPatchAction(...)
my_result.actions += EmailSomeone(...)
```
This may also solve https://github.com/coala/coala/issues/1021, as you can easily add different patches:
```
my_result.actions += ApplyPatchAction(...) # <-- patch variant 1
my_result.actions += ApplyPatchAction(...) # <-- patch variant 2
```
Some actions are implicitly added by coala, like the `IgnoreResultAction`.

CC @sils @jayvdb

**Comment by jayvdb** [comment]
Yes. We really need to slurp up pywikibot console functionality, which already has all this.

**Comment by satwikkansal** [comment]
@Makman2 
Just wanted to discuss an approach that I was thinking.

The issue involves allowing the bears to add custom actions apart from those provided by coala by default (`ApplyPatchAction`, `PrintMoreInfoAction`, `ShowPatchAction`, `IgnoreResultAction`)

The suggested approach involves
```sh
   Problem detected
        /  |   \  \
      /    |    \   \
    /      |     |    \ 
   A      B     C    D ...
```
Where A, B, C, D .. are Result.action instances.

I was wondering if instead of just collecting the actions in a list, we can also allow actions to be cascaded or combined based on logical if else conditions. So the bear can define an `ActionSequence` as 

```sh
   A    B       C       H        i
   |    |      |        |         |
   D    F     exit     exit     exit
   |   / \
exit  E   G
      |    |
    exit   exit
```
So users are initially given an option

```sh
|    | [INFO] SomeBear:
|    | Some problems in the code
|    | The following actions are applicable to this result:
|    |  0: Apply no further actions. (equivalent to exit)
|    |  1:   Action A
|    |  2:  Action B
|    |  3:  Action C
|    |  4:  Action H
|    |  5:  Action i
|    | Please enter the number of the action you want to execute.
```
And suppose if we chose Action B, then action F will be executed (automatically if it’s an implicit one, and after user’s consent if it’s an explicit one) and then depending on the results of F either E or G will be executed.

Each action returns a value, that would help the bears arrange actions the execution order of action in the form of a decision tree like structure. 
It'd still be possible to add/concatenate different action sequences like in the suggested method.
```py
Result.ActionSeq =  CoalaDefaultSeq(...)
Result.ActionSeq += XYZSeq(...)
```

A vague example for GitCommitBear :sweat_smile: 
```sh
|  | [INFO] Commit message does not contain issue url
|  | 1: Fix commit 
|  | 2: Do nothing
|  | Please enter the number of the action you want to execute.
```
Now suppose the users selects `1` which is a `FixCommitActionSequence`

Execute `VerifyCommitAuthorAction` (checks if the user is the author of the previous commit)
If he is not the author
```
|  | [INFO] You are not the author
|  | 1: Amend anyways
|  | 2: Email the commit owner
|  | 3: Skip
```
If he is the author OR he’s not but selected the “Amend anyways” option:

Execute the `InputValuesAction`
```sh
| | [INFO] Commit message does not contain issue url
| | The current commit message is :- 
| | Some Commit message with no issue url.
| | Please enter the new commit message

> Related to SOME_ISSUE_URL
```
Finally `AmendCommitAction` is executed and commit gets amended.

This is probably an over-construction of the solution. I know the example is very vague and the some of the actions could be combined into a single action but just wanted to convey the idea and know if it makes sense or not.

**Comment by Makman2** [comment]
Just ahead: ApplyPatchAction shall not be implicitly added by coala. Not all results have a patch. ShowPatchAction will have special behaviour, it gets added implicitly for each ApplyPatchAction the result contains.

Your approach sounds very interesting, and would enable very flexible behaviour that is more suited to the situation of the user. I think the easiest is to implement a field into `ResultAction`, called something like `next_action`. It's a list of actions the user can take after choosing after. 3 cases:
1. list empty: action done, do nothing more
2. one element inside list: execute `next_action` immediately
3. more than one element: provide a choice to the user which action to execute next

Maybe it should go into a new issue? Imo this can be implemented on top of [this approach](https://github.com/coala/coala/issues/3650#issue-203870459)

**Comment by satwikkansal** [comment]
Sounds good, we can have `next_action` field in the ResultAction.

> one element inside list: execute next_action immediately 

IMO, this should be the default behavior, but it should also be possible to mark actions as `explicit` or `implicit` (default one), because there maybe some actions like `EmailAction()` for which we may require user's consent before performing it.

**Comment by satwikkansal** [comment]
Is it possible to get assigned to this issue or is it blocked for GSoC reasons because once this functionality is implemented the multiple patch offering feature in https://github.com/coala/coala/issues/1021 can be solved just by adding some extra actions.

**Comment by Makman2** [comment]
> IMO, this should be the default behavior, but it should also be possible to mark actions as explicit or implicit (default one), because there maybe some actions like EmailAction() for which we may require user's consent before performing it.

This can't be default behaviour when having multiple actions defined for `next_action`. And if you want special confirmation, we should implement a `ConfirmationAction` doing this. As marking them `explicit` or `implicit` is totally messy. Also those would get meaningless when there is more than one action defined.

> Is it possible to get assigned to this issue or is it blocked for GSoC reasons because once this functionality is implemented, the "multiple patch suggestions" feature in #1021 can be implemented just by adding some extra actions.

Don't think this issue is part of a GSoC project, so feel free to pick it up together with the other one :+1:

**Comment by satwikkansal** [comment]
@Makman2 
Okay, so one approach I could think of is as follows:


We add a field "actions" inside `Result` class which is a list of ResultAction-instances.
```
Result.actions = [XYZAction(...)]
```
It's possible to set even multiple ones:

```
my_result = Result('Something went wrong')
my_result.actions += ApplyPatchAction(...)
my_result.actions += EmailSomeone(...)
```


- Now the question comes, How would a bear generate result with the possible actions?

We can add functionality for bears-specific `ResultAction` instances (defined in the implementation of the bear itself). But then, we may also prefer to re-use an already implemented `ResultAction` defined in some other bear (Example most of the bears may want to use the `InputValuesAction`.)

One solution may be to add all these custom ResultAction instances in `results/result_actions/custom` directory and in the implementation of these `ResultAction`s we can have a field called `BearsSupported` which is a set of bears on which the action is applicable. And we may allow every bear the flexibility to extend these ResultActions and define their own `is_applicable` criteria.

Also, inside bear's implementation, we may still support defining bear-specific ResultActions which are literally bear-specific and don't need to be added to `result-actions/custom-actions` directory because no bear will ever use them. (Example, ResultAction requires the linter/tool that the bear wraps to be installed)

The native bears can directly add these actions like
```py
yield Result.from_values(
                    origin
                    message,
                    ..
                    ..
                    actions = [ResultAction.custom_action.EmailAction,
                               ResultAction.custom_action.AmendCommitAction]
                    ..
                    ..
                    severity=RESULT_SEVERITY.INFO,
                    diffs=diffs)
```

For External bears, in the `@Linter` decorator, we can have an optional argument `result_actions` that gets intrinsically passed to the `process_output` function (like the `diff_distance` argument.), and the output processing functions then add the actions to the Results when yielding them.

One doubt that comes to my mind is should we also allow the bears to pass "params and their values" along with the list of ResultActions. (in case the bear wants to send some data that might be useful in executing the ResultAction)

- Now assuming that we have actions defined in the result field, how do we display them?

Currently, in the `ConsoleInteraction` module, in the `print_result` function, the user is prompted to chose any one of the applicable actions from the pre-defined `CLI_ACTIONS` list that contains OpenEditorAction, ApplyPatchAction,  PrintMoreInfoAction, ShowPatchAction(), etc.

imo, simply adding new actions to them should work.

```
cli_actions = CLI_ACTIONS + result.actions
```

**Comment by Makman2** [comment]
> Now the question comes, How would a bear generate result with the possible actions?

Like you've said, native bears have to instantiate a result and fill in information about the result (message, diffs, source ranges etc etc).

> One solution may be to add all these custom ResultAction instances in results/result_actions/custom directory and in the implementation of these ResultActions we can have a field called BearsSupported which is a set of bears on which the action is applicable. And we may allow every bear the flexibility to extend these ResultActions and define their own is_applicable criteria.

This would make `coala` dependent on `coala-bears` in principle, let's not do this. The idea that this is the pure core, nothing more (heh that was a rhyme :3)

> For External bears, in the @Linter decorator, we can have an optional argument result_actions that gets intrinsically passed to the process_output function (like the diff_distance argument.), and the output processing functions then add the actions to the Results when yielding them.

I would just improve the current parsing algorithms to use suited actions.

> Currently, in the ConsoleInteraction module, in the print_result function, the user is prompted to chose any one of the applicable actions from the pre-defined CLI_ACTIONS list that contains OpenEditorAction, ApplyPatchAction, PrintMoreInfoAction, ShowPatchAction(), etc.
> imo, simply adding new actions to them should work.
> `cli_actions = CLI_ACTIONS + result.actions`

Exactly what I had in mind ;)

**Comment by satwikkansal** [comment]
> This would make coala dependent on coala-bears in principle, let's not do this. The idea that this is the pure core, nothing more (heh that was a rhyme :3)

Yeah, makes sense :)  So should we allow the `ResultAction`'s implementation to be defined in the core or the bear's implementation. 
Because if we add in coala-core, then if a bear writer has to add a custom action, he would first have to implementation the `ResultAction` in the core and then use it in the bear. But if we allow custom `ResultAction` to be defined in bear's implementation, we'll also need to develop a way to reuse these Result actions across multiple bears.

**Comment by Makman2** [comment]
>  But if we allow custom ResultAction to be defined in bear's implementation, we'll also need to develop a way to reuse these Result actions across multiple bears.
  
 We merge these into the core if they are useful, otherwise the user can't implement actions which have their own very different behaviour. If the behaviour is implementable by other users, actions can be shared by sharing the module.

**Comment by satwikkansal** [comment]
Makes sense. Should I come up with more detailed design or simply start working on the implementation? :sweat_smile:

**Comment by Makman2** [comment]
I think the direction and API of the issue is clear, imo you can start ;)

**Comment by satwikkansal** [comment]
Okay, I've started working on this.

> For External bears, in the @Linter decorator, we can have an optional argument result_actions that gets intrinsically passed to the process_output function (like the diff_distance argument.), and the output processing functions then add the actions to the Results when yielding them.

> > I would just improve the current parsing algorithms to use suited actions.

@Makman2 I think I didn't understand this ^^. How are parsing algorithms supposed to determine which actions are suited?  Don't the actions need to be supplied by the bear writer? :)

**Comment by Makman2** [comment]
Indeed, though an `@linter`-bear takes over the result-generation itself, so in most cases we can offer fitting actions. For example: If someone specifies `corrected` output-format, there will be no other actions than `ApplyPatchAction`s (and `ShowPatchAction`s) normally. The same principle applies to other output-formats I think :+1:
If the need arises for an `@linter` writer to adjust actions, we should take this to a new issue imo :+1:

**Comment by satwikkansal** [comment]
I'm confused regarding handling these NATIVE_ACTIONS (ShowPatchAction, OpenEditorAction, etc which coala adds implicitly to a result) for these new use cases of custom actions and next actions. A few design related questions that I've are:
- Should we allow users to pass native actions like OpenEditorAction in the actions field.
- If yes, then how do we handle multiple occurrences of such actions in result.actions
- Should the users be allowed to define next_actions for these Native actions, and if yes, how are they going to do so.
- Or should we treat the native actions and the custom actions alike (This will shift the design overhead to bear-writer which is not good imo) 

I think we'll need to design it properly. We will also need to handle the multiple instances of same ResultAction class, because we may not want the user to see more than one "OpenEditorActions"

One approach to remove duplicate actions of same class type: In case of duplicacy, the action having non-empty next_actions field is preferred. If the duplicity still presists, the most recent action is given preference.

**Comment by Makman2** [comment]
That's actually a good question.

> Should the users be allowed to define next_actions for these Native actions, and if yes, how are they going to do so.

That's one of the main points why we should expose native actions. That means we treat them like normal
actions.

If a user defines multiple actions which "don't make sense" (like multiple `OpenEditorActions`), we have to filter that properly.

Or:
Maybe we go away from the concept of "native actions" at all, and try to group actions better. For example:
`ApplyPatchAction` won't be used directly now by the user. Instead there's a `PatchAction`, which defines for `next_actions` two alternatives implicitly: `ShowPatchAction` and `ApplyPatchAction`.

**Comment by satwikkansal** [comment]
A simple solution maybe:
- We allow the users to use native actions and custom actions alike.
- And we still maintain the default implicit behavior that we have currently.
- In case of multiple native actions provided, we simply ignore the default ones and use only the user-provided native action (So if the user wants to add `next_actions` to the native actions, he can do it this way)
- And yes, we filter out other kind of actions that don't make sense together based on some criteria (like multiple `OpenEditorActions`). Also, if two functions have same class but different `next_actions`, I think we should display both because now they may represent different sequences of actions (but for this, we need to figure out a way to make the user realize how these actions are different may lead to different outcomes)

>For example: ApplyPatchAction won't be used directly now by the user. Instead there's a PatchAction, which defines for next_actions two alternatives implicitly: ShowPatchAction and ApplyPatchAction.

Grouping those two actions sounds like a good idea to me. But, in general, allowing users to pass patch related actions would require a lot of refactoring. To offer multiple `PatchAction`, we may want to move the `Diff` out of the `Result` class to `PatchAction` class, which means we'll have to explicitly pass the Diffs along with the Result elsewhere. So imo, we should solve this specific case in a separate issue, and specifically, in https://github.com/coala/coala/issues/1021
PS: As it's a GSoC project as well, there maybe other approaches which we may want to consider :)

**Comment by satwikkansal** [comment]
Also, should we wait for https://github.com/coala/coala/issues/4114 to be resolved first, or should I continue working on this issue

**Comment by Makman2** [comment]
No you can continue working on that, #4114 can be done later.

> But, in general, allowing users to pass patch related actions would require a lot of refactoring.

Not really. Everywhere where a user sets `Result(..., diff=...)`, he does instead `my_result.actions += ApplyPatchAction(my_diff)`. It's quantity, not quality in this case ;D

> To offer multiple PatchAction, we may want to move the Diff out of the Result class to PatchAction class

Exactly what I wanted :) Actually the diff/patch is only useful when somebody has the possibility to apply it, so it's bound to an action, not to a result (I mean not directly).

And generally I agree with your approach of how to handle user-passed native actions, we just need to do proper filtering.

**Comment by satwikkansal** [comment]
> And generally I agree with your approach of how to handle user-passed native actions, we just need to do proper filtering.

@Makman2 For filtering, how about we add a class variable called `allow_multiple` to the ResultActions, it is `False` by default, and if it's `True` then only we'll allow multiple actions of the same type.

**Comment by Makman2** [comment]
Hm not sure if I want to shift that design-overhead to the writer of a `ResultAction`. Not that much time to think that through completely now, but I'll be back ;D

**Comment by satwikkansal** [comment]
No problem :)

Also, should we consider moving the  "filtering based on applicablity of the actions" part from `ConsoleInteraction` to `Result.action` field (I mean only applicable `ResultAction`s are returned when we use the property `my_result.actions`.

**Comment by Makman2** [comment]
deprecate the `Diff` feature by emitting a warning and implicitly set an action via `result.action`.

> Another question, should we consider moving the "filtering based on applicablity of the actions" part from ConsoleInteraction to Result.action field (I mean only applicable ResultActions are returned when we use the property my_result.actions.

phew good question... like you want :)

Actually consider what you think is best, as we have many possibilities to solve this issue^^

**Comment by satwikkansal** [comment]
@Makman2 For deprecating `Diff`,
1. Group `ShowPatchAction` and `ApplyPatchAction` into `PatchAction` (a ShowPatchAction with `next_action` set to ApplyPatchAction)
2. Each `PatchAction` is initialized with exactly one diff, but one Result can contain multiple PatchActions (atleast one).
3. PatchAction is implicitly added to `result.actions` when result is initialized

> deprecate the Diff feature by emitting a warning and implicitly

Would it be sufficient to add  a logging.warning when non-empty `Diff` object is passed to any constructor of `Result` class.

**Comment by Makman2** [comment]
> PatchAction is implicitly added to result.actions when result is initialized

No, not every result has a patch. The user adds it himself.

> Would it be sufficient to add a logging.warning when non-empty Diff object is passed to any constructor of Result class.

Also add a `PatchAction` to the new `result.actions`, so diffs do still work for a while :+1:

**Comment by Vedarth** [comment]
@Makman2 This issue is also related to the GSoC project that I am interested in. Can you please assign it to me, as this has been inactive for a long time. I would love to work on it.

**Comment by Naveenaidu** [comment]
@Makman2 Can i please be assigned to this issue.

**Comment by jayvdb** [comment]
I suggest you both try some smaller issues before tackling something like this.

**Comment by Makman2** [comment]
Agreeing with @jayvdb, this should be designed out properly, and can be part of gsoc itself.

**Comment by satwikkansal** [comment]
Unassigning myself due to inactivity :sweat_smile: If anyone is interested in taking this up for GSoC, the code and discussions on the PR https://github.com/coala/coala/pull/4112 might be helpful.

**Comment by Vedarth** [comment]
@satwikkansal thanks for your help. It is very useful.
